### DISCUSSION TITLE
API use on shared server and daily limit

### DISCUSSION THREAD

**Discussion by ludosevilla** [open-meteo/open-meteo#853] [Q&A]
Hello,

I use this powerful and wonderful API on a shared server.

Today, when using the API, it returns error=1 and the message "Daily API request limit exceeded. Please try again tomorrow." , and it was the first request of the day...

I don't know how are counted the number of requests, but if it is per IP, I guess there could be an issue with shared server, as the one I am using, all the applications using the open-meteo API having the same IP, even if their owner are totally different.

I guess that an other service (not mine) have exceeded the requests limit...
In that case, what is the solution, except a paying plan, to keep using the API?
It is not very fair that an unfair use by another will penalize others..

My service is free, no ads (and source available on github) and make as a maximum of 10 or 20 requests / week..
So paying for the API is not a solution as the lowest price is 30 euros/month..

Is there a possibility to get a private API key to differenciate the users?
If not, maybe a possibility to credit an account, and for example each X request would cost 1 cent ?

Thanks

**Comment by rany2** [comment]
I made a different issue which I think is similar to what you're asking for: https://github.com/open-meteo/open-meteo/discussions/843

I believe that the reason you're exhausting your quota is similar to what my users are facing.

**Reply by ludosevilla** [reply]
Yes, I think so. I hope there will be soon a solution... I would be sad to stop using open-meteo because of this (unfair) limitation.

**Reply by rany2** [reply]
It's not necessarily unfair, using Open-Meteo is a privilege not a right. We're not entitled to access.

At any rate, I've worked around this by self-hosting Open-Meteo. I needed around ~1.5TB of disk space and it's really straightforward. The docs are great.

**Reply by ludosevilla** [reply]
When I say "unfair", I don't mean that the limitation is inself unfair (as you said, it s a privilege, not a right), but unfair in the meaning that people that make a "fair" use of open-meteo will face the consequence of people makeing an "unfair" use of the system.

**Reply by FleetAdmiralJakob** [reply]
> When I say "unfair", I don't mean that the limitation is inself unfair (as you said, it s a privilege, not a right), but unfair in the meaning that people that make a "fair" use of open-meteo will face the consequence of people makeing an "unfair" use of the system.

Hi, I'm hitting the rate limit too with using Vercel without any more requests from my side, see: https://github.com/open-meteo/open-meteo/issues/811

But I dont think that the people that are causing my issue do unfair stuff. It's just the nature of the current system in which someone will always hit the limit and this will break also other apps.

And (I dont have numbers here), but it also dont have to mean that some people are abusing the system. The problem can also be that too many people are using the same IP adress at the same time.

But I agree with you @ludosevilla and @rany2. Im really thankful that this project exists and that is offering a free plan thank you for that @patrick-zippenfenig. 

To solve the problem: I have to admit that it's really hard to find a better option that is sustainable at the same time.

**Reply by ludosevilla** [reply]
> > When I say "unfair", I don't mean that the limitation is inself unfair (as you said, it s a privilege, not a right), but unfair in the meaning that people that make a "fair" use of open-meteo will face the consequence of people makeing an "unfair" use of the system.
> 
> Hi, I'm hitting the rate limit too with using Vercel without any more requests from my side, see: #811
> 
> But I dont think that the people that are causing my issue do unfair stuff. It's just the nature of the current system in which someone will always hit the limit and this will break also other apps.
> 
> And (I dont have numbers here), but it also dont have to mean that some people are abusing the system. The problem can also be that too many people are using the same IP adress at the same time.
> 
> But I agree with you @ludosevilla and @rany2. Im really thankful that this project exists and that is offering a free plan thank you for that @patrick-zippenfenig.
> 
> It's really hard to find a better option that is sustainable at the same time.

Well, you're right! Probably the word "unfair" is not the correct one! And thanks again to open-meteo for existing!

**Comment by patrick-zippenfenig** [comment]
Hi,
thanks for the report. Yes, in the recent weeks more and more people reported this issue for shared services like vercel, cf pages, etc.

Offering API keys for non-commercial use is an option, but also offers large potential for misuse. Getting multiple API keys and effectively circumventing any rate limit is easy. The existing IP limit can also be circumvented, but using multiple IP addresses is actually more complicated.

It would also require multiple days to develop the sign-up system, databases, synchronisation across multiple datacenters, etc. For paying customers that cross-finance the free API, I use different systems that would be unsustainable (= $$$ expensive) to offer for free. A simple solution would be to offer a "non-commercial" API key at low rates like $2-3 per month, but this can lead to more support effort and can be abused by commercial users.....

Other ideas? Any way to authenticate users more reliably to prevent multi-accounts? GitHub SSO only? Any software solutions out there to automate this setup (My favourite would be a SvelteKit based signup system on CF workers with a CF database)?

**Reply by rany2** [reply]
> Any way to authenticate users more reliably to prevent multi-accounts? GitHub SSO only?

GitHub SSO is a really bad idea as one could create an account anonymously and it's not against TOS. I think that if SSO was used then Facebook or Google should be the only ones allowed as they do actually crack down on fake accounts. I do think that Facebook is more strict though.



> For paying customers that cross-finance the free API, I use different systems that would be unsustainable (= $$$ expensive) to offer for free. A simple solution would be to offer a "non-commercial" API key at low rates like $2-3 per month, but this can lead to more support effort and can be abused by commercial users.....

I think the best approach if you don't want to go the SSO route is to create a pay-as-you-go option were one API request is something like 0.0002 EUR. This way it's always cheaper to opt for the commercial option, by the time you reach 140,000 requests; it becomes more cost effective to use the commercial option. Optionally, you could make this another commercial tier. You could play around with the numbers of course, but I feel like it's the best solution to protect against abuse.

**Reply by rany2** [reply]
> I think the best approach if you don't want to go the SSO route is to create a pay-as-you-go option were one API request is something like 0.0002 EUR.

Just to clarify a bit more, you'll need to credit your account a minimum of say 5 EUR. So it's not actually PAYG, it's more like a bus card system.

**Reply by rany2** [reply]
> GitHub SSO is a really bad idea as one could create an account anonymously and it's not against TOS. I think that if SSO was used then Facebook or Google should be the only ones allowed as they do actually crack down on fake accounts. I do think that Facebook is more strict though.

Besides Facebook and Google SSO, you could also treat university emails as privileged. There exists this repository that helps you detect if the college email could be accepted: https://github.com/JetBrains/swot but I think their criteria is too harsh (banning universities where students keep their emails after graduation)

**Reply by FleetAdmiralJakob** [reply]
Hi, I'm currently a student but I dont have a student email because I'm still in middle school and not at university but can be checked if my GH Account is in the GH Student Program or JetBrains Account is in the JetBrains Student Program?

Because there you have to authenticate that you are a real person.

This would take the overhead of checking if the people are real away from open meteo because GH and JB are doing it for you.

**Reply by rany2** [reply]
Not sure about how complex that would be because he'd need to figure out how to partner with GitHub: https://github.com/edu/partners and the participation seems opaque (there's just a survey you fill out and you'll need to wait afterwards). 

Anyway the JetBrains Student Program is only meant for university students, it says as much in the ReadMe:

> NOTE: If an organization provides primary or secondary education only (i.e., no high or higher education programs), it will not be included in the list. Primary and secondary school students do not usually need access to professional developer tools, and if they learn some programming, we are glad to offer them Community versions of the tools, such as PyCharm Community Edition, to use which are free to everybody.

**Reply by FleetAdmiralJakob** [reply]
> Anyway the JetBrains Student Program is only meant for university students, it says as much in the ReadMe:
> 
> > NOTE: If an organization provides primary or secondary education only (i.e., no high or higher education programs), it will not be included in the list. Primary and secondary school students do not usually need access to professional developer tools, and if they learn some programming, we are glad to offer them Community versions of the tools, such as PyCharm Community Edition, to use which are free to everybody.

Oh then I somehow dribbled the system  🤣

**Reply by FleetAdmiralJakob** [reply]
One, question, why is it so much harder and costs more to rate limit via api key instead of ip.

Dont you need for both a db in which the ip/api keys are stored?

**Reply by patrick-zippenfenig** [reply]
API keys require a signup-form, email verification, a **central** database with permanent storage and synchronisation with all API nodes. Limits by IP can be as simple as a hash-table in the application memory

**Reply by FleetAdmiralJakob** [reply]
I wonder if IPv6 can help here. Vercel probably reuses a lot of IPv4 adresses which cause the ratelimiting then. Maybe with checking IPv6 there are less such errors, but I'm not sure. Just an idea.

**Reply by FleetAdmiralJakob** [reply]
@patrick-zippenfenig And what about using a service like unkey? I dont know how many users open-meteo has but for me, it seems relatively cheap at the price of 25 bucks for 1m api keys

https://www.unkey.com/pricing

**Reply by SebSwan** [reply]
The free tiers is really generous. It look like a no-solution for a no-problem, if a client need more api call, he take a premium subscription or find a solution to handle a different IP for calling the api. Time and resources of @patrick-zippenfenig is limited, i prefer a focus on more valuable features.
